<img src="uc3m-logo.jpg" alt="UC3M logo" width="200"/>

# Second Homework: Network Optimization and Non-linear Models (Topics 2, 3)
## BsC in Data Science and Engineering - Optimization and Analytics (2022-2023)

Author: Rodrigo Oliver Coimbra (NIA: 100451788)

## 1. Introduction (network optimization)

Companies and governments are facing increasingly complex and intricated problems that require rigorous assessment and monitoring, something that can be be accomplished through the use of mathemtical tools such as, for example, linear programming techniques. 

In today's world, competitiveness and productivity are of paramount importance when comparing businesses, estimating their revenue and determining their solvency. Many companies and governments are lagging behind in the implementation of rational, quantitative solutions that would otherwise enable them to enact the necessary measures to achieve their objectives; we need not to look any further than the report *Key factors behind productivity trends in euro area countries* released by the European Central Bank which states that:

> Productivity, defined broadly as efficiency in production, plays a key role in the economic resilience and social welfare of countries. Productivity growth influences the economy in important ways, affecting key variables such as output, employment and wages. Productivity is also relevant for monetary policy as it is a fundamental determinant of potential output growth and the natural rate of interest and, therefore, of the monetary policy space needed to deliver price stability over the medium term. As such, changes in productivity can influence the transmission mechanism of monetary policy and should be closely monitored.

and:

> [...] aggregate labour productivity growth has been trending downwards for decades, both in the euro area and in other major economies. Average annual growth in labour productivity – measured as real GDP per hour worked – in those euro area countries that have sufficiently long time series has continuously declined from about 7% in the 1960s to just 1% since the early 2000s [...]

Project management is not the only ingredient in the recipe for productivity and efficiency but it is nonetheless an important factor that has the potential of drastically improving these indicators and eliminating removable bottlenecks caused by suboptimal planning policies.

An article published by the Harvard Business Review Editors named *The Four Phases of Project Managament* argues that:

> [...] Many projects fail either because they bite off more than they can chew and thus grossly underestimate time and money or because a significant part of the problem has been overlooked. [...] Ask, "What will have to be done in order to accomplish X?". [...] Estimate how long it will take to complete these tasks and how much tey will cost in terms of dollars and person-hours. As a result of your thoughtful planning [...] you'll also have a good idea of how long the project will take.

HBR further elaborate that:

> [...] most projects come with fixed beginning and end dates, regardless of available resources. To create a realistic schedule within those constraints, work backward from any drop-dead deadlines you know about- that is, dates that cannot be changed- to see when your deliverables must be ready.

Business ought not to indulge in wishful thinking but instead focus on a rational allocation of capital, human resources and, crucially, time. For a comprehensive review of the impact of project management on productivity, refer to this report, published by the Association for Project Management, a British chartered professional organization.

Another article, by EY's Oceania Business Transformation Leader, Loretta Hudson, states that:

> Our research shows that miners that invest in the right project management capabilities and toolsets can reap between 1% and 30% of project value.

Bertsekas' book briefly mentions an example corresponding to the subject of Project Management (vide p.54) but does not elaborate any further on the matter. After doing some research on the subject, I discovered that network linear programming may be used to solve problems related with the Critical Path Method (CPM) and Program Evaluation and Review Technique (PERT).

While CPM/PERT and its derivative models do not cover or take into account all the factors involved in project management, they do provide useful insight into the dynamics of the problem to be addressed; accordingly, CPM was developed at DuPont and Remigton Rand in 1950s and PERT was designed by the United States Navy at the same period. These techniques represented a significant leap into modern project management with respect to other tools, such as Gantt charts.



## 2. Network optimization (with discrete variables)

### 2.0 Model explanation 

Managers and business owners face the classical problem of allocating resources, creating a sensible work flow and determining deadlines. Some common questions posed are: *When will the project be completed?*, *Which activities are critical for the correct completion of the project within a certain deadline?*, *Do we have enough resources to correctly complete the project?*, *Is the project running on schedule?*, *Are we using more resources than those budgeted?*, *Can we diminish the project duration with the lowest cost possible and how to so?*, etc.

In this network model we will focusing on the completion of a project within a certain specified deadline using the least possible cost, which would otherwise take much longer or extend beyond the deadline - deal-braker for most businesses.

For very simple projects, a heuristic approach by inspection may be used or calculated by hand using an appropiate algorithm (apart from the linear programming model, there are innumerable specialized algorithms for solving particular network problems). These approaches (the heuristic or hand-calculated), however, have serious limitations as they cannot be accomodate for larger projects, which mean that their solutions are far from being optimal and cannot therefore, be deemed to be fit for optimization.

#### 2.0.0 General conditions (Network Crashing with CPM/PERM)

In this particular model we will be using information organized according to CPM for obtaining the expected completion time of events and the usage of the least monetary resources to reduce the duration of the project, this is called **network crashing**.

We need a set of events and activities; events occur at a certain point in time and activities have a normal duration, normal cost, reduced (crashed) duration and reduced (crashed) cost.

It is clear that some activities depend on the successful completion of others and cannot be iniatiated earlier. We will be using events to connect and codify these dependency relationships between activities. Dummy activities may be added to ensure the logical value of the project is preserved. Bear in mind that dummy activities have no normal duration and thus cannot be reduced (crashed).

#### 2.0.1 Specific conditions

The network below corresponds to a CPM project network; this is extremely flexible as it can represent either physical or abstract events and activities for any kind of activities for any scale: small, medium or large endeavors. 

In this case we focus on the CPM project network of a given company that it is interested in knowing which would be the least amount of money that would be invested in order to finish the project is less time, namely, before the deadline that is specified and, perhaps most importantly, knowing in which activities to do so and in what amount.

The problem start with the initial state 0 and ends with the final states 21. There are in total 22 events (nodes) and 28 activities (arcs), which means that the incidence matrix of this network has 616 entries.

<img src="network_diagram2.svg" alt="Project management network"/>

As explained in the previous subsection, all activities have a name, normal duration, normal cost, reduced (crashed) duration and reduced (crashed) cost. This information is given in the following table for this specific project:

| Activities | Predecessor(s) | Normal time (in days) | Crash time (in days) | Normal cost (in USD) | Crash cost (in USD) |
|------------|----------------|-----------------------|----------------------|----------------------|---------------------|
| A          | -              | 5                     | 2                    | 1,000                | 1,500               |
| B          | A              | 2                     | 1                    | 1,500                | 2,000               |
| C          | B              | 10                    | 3                    | 5,500                | 7,000               |
| D          | A              | 3                     | 1                    | 250                  | 500                 |
| E          | D              | 7                     | 2                    | 400                  | 650                 |
| F          | E, Dummy 1     | 5                     | 1                    | 500                  | 750                 |
| G          | E, Dummy 1     | 6                     | 2                    | 700                  | 800                 |
| H          | F              | 2                     | 1                    | 500                  | 900                 |
| I          | G              | 3                     | 1                    | 100                  | 300                 |
| J          | G              | 15                    | 4                    | 3,500                | 5,500               |
| K          | H              | 12                    | 3                    | 3,000                | 4,750               |
| L          | I              | 4                     | 1                    | 600                  | 800                 |
| M          | J              | 8                     | 3                    | 1,000                | 1,400               |
| N          | K, L, M        | 5                     | 2                    | 700                  | 1,000               |
| O          | K, L, M        | 10                    | 1                    | 2,800                | 3,100               |
| P          | N              | 20                    | 3                    | 6,000                | 6,900               |
| Q          | O              | 5                     | 2                    | 700                  | 800                 |
| R          | O              | 2                     | 1                    | 300                  | 450                 |
| S          | P              | 3                     | 1                    | 200                  | 300                 |
| T          | P              | 15                    | 3                    | 4,000                | 4,700               |
| U          | Q              | 21                    | 5                    | 6,000                | 8,500               |
| V          | R              | 30                    | 8                    | 10,500               | 16,000              |
| W          | S, Y           | 10                    | 3                    | 3,800                | 4,000               |
| X          | W, Dummy 2     | 3                     | 1                    | 200                  | 250                 |
| Y          | C              | 2                     | 1                    | 3,000                | 3,500               |
| Z          | T              | 5                     | 2                    | 1,500                | 2,000               |
| Dummy 1    | C              | 0                     | 0                    | 0                    | 0                   |
| Dummy 2    | U, V, Z        | 0                     | 0                    | 0                    | 0                   |

### 2.1 General formulation

* **Decision variables**:
    * **X**: Point in time (in this case days) when event $i$ takes place, with $0 \leq i \leq 21$. Therefore $V = 22$.
    * **y**: Reduction in time units (in this case days) that activity $j$ is subject to, with $0 \leq j \leq 27$. Therefore $A = 28$.

* **Objective function**: <span style="color:red">Minimize</span> the cost of reducing the projct duration (crashing) per the product between the reduction (crash) cost per activity $j$ per time unit.

$$\min_{y_j}\sum_{j=0}^{A-1}c_j·y_j$$

* **Constraints**:
    * **Crash time constraints**: the crash time for a given activity j must be lesser or equal to the difference between the normal and crashed times for that activity. ($n_j$ stands for the normal duration of activity $j$ and $r_j$ stands for the reduced duration of activity $j$).
    
    $$y_j \leq n_j - r_j$$

    * **Network (limit) constraints**: the time where a certain event $i$ (node $i$)
    is reached (with respect with the initial event time of 0) is always greater or equal to
    the result of the normal duration of activity $j$ that is related with the occurrence of 
    the event, minus the crashing time of the said activity $j$ plus the occurrence time of
    the predecessor event $k$ (node $k$) where the activity $j$ originates from.
    
    $$x_i \leq n_j - y_j + x_k \quad \forall \quad 0 \leq i \leq V, \quad j = \{k, i\}$$

    * **Project completion constraints**: the time occurence of the initial event is equal to 0 and that the conclusion date of the project must be lesser or equal than the deadline that is specified.

    $$x_0 = 0$$
    $$x_{V-1} = d, \quad d \triangleq \text{deadline}$$

    * **Non-negativity and discrete value of the solution**: the value of either $x_i$ or $y_j$ must always be larger or equal to zero, as they represent amount of time.
    $$x_i \in \mathbb{Z}^{+} \cup \{0\} \quad \forall \quad 0 \leq i \leq V-1$$
    $$y_j \in \mathbb{Z}^{+} \cup \{0\} \quad \forall \quad 0 \leq j \leq A-1$$

    With respect to this last constraint, we **integrality theorem** states that:

    > Project completion constraints establish that the time occurence of the initial event is equal to 0 and that the conclusion date of the project must be lesser or equal than the deadline that is specified.

    given that the network crashing problem is a subvariant of the minimum cost network flow problem, we can apply this nice result to our problem, which means that vanilla linear programming with the simplex algorithm that is employed for continuous problems can be used without the need for additional methods or the burden of more complex algorithms for solving discrete problems. We will be checking that this theorem indeed holds by intentionally creating the variables and setting them with domain in the reals and then checking that the result is nonetheless discrete.

### 2.2 Python implementation of network concepts

In [1]:
# We import the libraries that are needed
# for solving linear programming problems
import pyomo.environ as pyo
from pyomo.opt import SolverFactory

# For better data manipulation
# we import Numpy, the Numeric Python library
import numpy as np

# For performance analysis purposes
from time import process_time

# For saving the results returned by the solver
from pandas import Series, DataFrame

# For better code readability and maintanability
from typing import Final, List, Dict
from dataclasses import dataclass

In [2]:
class Activity:
    """
    This class defines the activities that make up the project to be crashed (optimized),
    each activity has a name (to make it more human-readable), a start event, an end event,
    a normal duration, a crashed duration, the normal cost that it represents and the cost
    that crashing would entail.
    """
    
    def __init__(self, name: str, start: int, end: int, normal_time: int, crash_time: int, normal_cost: int, crash_cost: int):
        self.name = name
        self.start = start
        self.end = end
        self.normal_time = normal_time
        self.crash_time = crash_time
        self.normal_cost = normal_cost
        self.crash_cost = crash_cost
        
        if normal_cost == 0 or crash_time == 0:
            self.C = 0
        else:
            #self.C = int((crash_cost - normal_cost)/(normal_time - crash_time))
            self.C = (crash_cost - normal_cost)/(normal_time - crash_time)

def generateA (A: np.ndarray, start_nodes: List[int], end_nodes: List[int]) -> None:
    """
    This function takes an empty matrix A, a list of 
    starting nodes and a list of ending ones; it converts
    it into an incidence matrix of the activities (arcs)
    and events (nodes)
    """
    
    count = 0
    for i in zip(start_nodes,end_nodes):
        A[i[0]][count] = 1
        A[i[1]][count] = -1
        count+=1
    
    return

def startNodes(alist: List[int]) -> List[int]:
    """Returns the list of starting events (nodes)"""
    return [(a.start) for a in alist]

def endNodes(alist: List[int]) -> List[int]:
    """Returns the list of ending events (nodes)"""
    return [(a.end) for a in alist]

def diffTime(alist: List[int]) -> List[int]:
    """Returns the time difference between
    the normal duration and the crash one"""
    return [(a.normal_time - a.crash_time) for a in alist]

def cValues(alist: List[int]) -> List[int]:
    """Returns the cost-to-time ratio"""
    return [a.C for a in alist]

def normalTime(alist: List[int]) -> List[int]: 
    """Returns the normal time duration by activity"""
    return [a.normal_time for a in alist]

def networkDict(alist: List[int], nodes: int, arcs: int, M: np.ndarray) -> Dict:
    """
    Returns a dictionary containing the following information:
    (i,j):(a,b) where (i,j) are the rows and columns of the incidence
    matrix - respectively - and (a,b) represent the start and end
    nodes of a give activity.
    """
    
    temp = {}
    res = {}
    for i, e in enumerate(alist):
        temp[i] = (e.start, e.end)
    for i in range(nodes):
        for j in range(arcs):
            if M[i][j] == -1:
                res[(i,j)] = temp[j]
    return res

def uniqueEndNodes(alist: List[int]) -> int:
    """Returns a list of unique end nodes"""
    return len(set(endNodes(alist)))

def numberNodes(alist: List[int]) -> int:
    """Returns the number of unique nodes in the model"""
    return len(set(endNodes(alist))) + 1

### 2.3 Data creation and manipulation

In [3]:
# We create the activity arcs between the
# event nodes with all the necessary information

activity_list = [
    Activity('A', 0, 1, 5, 2, 1_000, 1_500),
    Activity('B', 1, 2, 2, 1, 1_500, 2_000),
    Activity('C', 2, 3, 10, 3, 5_500, 7_000),
    Activity('D', 1, 4, 3, 1, 250, 500),
    Activity('E', 4, 5, 7, 2, 400, 650),
    Activity('F', 5, 6, 5, 1, 500, 750),
    Activity('G', 5, 7, 6, 2, 700, 800),
    Activity('H', 6, 8, 2, 1, 500, 900),
    Activity('I', 7, 9, 3, 1, 100, 300),
    Activity('J', 7, 10, 15, 4, 3_500, 5_500),
    Activity('K', 8, 11, 12, 3, 3_000, 4_750),
    Activity('L', 9, 11, 4, 1, 600, 800),
    Activity('M', 10, 11, 8, 3, 1_000, 1_400),
    Activity('N', 11, 12, 5, 2, 700, 1_000),
    Activity('O', 11, 13, 10, 1, 2_800, 3_100),
    Activity('P', 12, 14, 20, 3, 6_000, 6_900),
    Activity('Q', 13, 18, 5, 2, 700, 800),
    Activity('R', 13, 19, 2, 1, 300, 450),
    Activity('S', 14, 15, 3, 1, 200, 300),
    Activity('T', 14, 16, 15, 3, 4_000, 4_700),
    Activity('U', 18, 20, 21, 5, 6_000, 8_500),
    Activity('V', 19, 20, 30, 8, 10_500, 16_000),
    Activity('W', 15, 17, 10, 3, 3_800, 4_000),
    Activity('X', 17, 21, 3, 1, 200, 250),
    Activity('Y', 3, 15, 2, 1, 3_000, 3_500),
    Activity('Z', 16, 20, 5, 2, 1_500, 2_000),
    Activity('Dummy 1', 3, 5, 0, 0, 0, 0),
    Activity('Dummy 2', 20, 17, 0, 0, 0, 0)
]

# Define three important constants:
# the number of events (nodes); the number of
# activities (arcs); and the deadline in days
# of the project
V: Final = numberNodes(activity_list)
A: Final = len(activity_list)
DEADLINE: Final = 60

# Check that E and A are equal
# to their theoretical value
assert V == 22
assert A == 28

# Obtain two lists: one with the 
# starting nodes and another one
# with the ending ones
start_nodes = startNodes(activity_list)
end_nodes = endNodes(activity_list)

# Create a E times A incidence matrix
# in this case it has 22*28 = 616 entries
M = np.zeros((V, len(end_nodes)))
generateA(M, start_nodes, end_nodes)

# Check that M has the correct shape
assert M.shape == (V, A)

# Create a dictionary with all the necessary 
# information to implement the network constraints
activity_dict = networkDict(activity_list, V, A, M)

# Check that number of elements in the dictionary
# correspond to the number of activities
assert len(activity_dict) == A

# Print the dictionary
print(f"{activity_dict=}")

activity_dict={(1, 0): (0, 1), (2, 1): (1, 2), (3, 2): (2, 3), (4, 3): (1, 4), (5, 4): (4, 5), (5, 26): (3, 5), (6, 5): (5, 6), (7, 6): (5, 7), (8, 7): (6, 8), (9, 8): (7, 9), (10, 9): (7, 10), (11, 10): (8, 11), (11, 11): (9, 11), (11, 12): (10, 11), (12, 13): (11, 12), (13, 14): (11, 13), (14, 15): (12, 14), (15, 18): (14, 15), (15, 24): (3, 15), (16, 19): (14, 16), (17, 22): (15, 17), (17, 27): (20, 17), (18, 16): (13, 18), (19, 17): (13, 19), (20, 20): (18, 20), (20, 21): (19, 20), (20, 25): (16, 20), (21, 23): (17, 21)}


### 2.4 Network Linear Programming implementation

#### 2.4.1 Problem data definition: sets and parameters

In [4]:
# We create the Pyomo model to be used
network_model = pyo.ConcreteModel()

# Create two lists: one for the event labels
# and another one for the activities labels
# to be used a Sets for the model
event_labels = [v for v in range(V)]
activity_labels = [a for a in range(A)]

# Initialize both of these lists to the model,
# with the events corresponding to the i dimension
# and the activities corresponding to the j dimension
network_model.i = pyo.Set(initialize=event_labels)
network_model.j = pyo.Set(initialize=activity_labels)

# Fetch the c values for all activities in the model
c = cValues(activity_list)
# Initialize these values into the model as parameters
network_model.costs = pyo.Param(network_model.j, initialize=c, doc="Cost in monetary units per activity per time unit")

# Fetch the normal time values for all activities in the model
normal_time = normalTime(activity_list)
# Initialize these values into the model as parameters
network_model.normaltime = pyo.Param(network_model.j, initialize=normal_time, doc="Normal time in days for the completion of a given activity")

# Fetch the difference between normal and crashed times
# for all activities in the model
diff_time = diffTime(activity_list)
# Initialize these values into the model as parameters
network_model.timediff = pyo.Param(network_model.j, initialize=diff_time, doc="Time difference in days between the normal time and the crash time")

#### 2.4.2 Define the variables

In [5]:
# Define and add to the model the x and y variables,
# living in the i and j dimensions, respectively
network_model.x = pyo.Var(network_model.i, within=pyo.NonNegativeReals)
network_model.y = pyo.Var(network_model.j, within=pyo.NonNegativeReals)

# Print the variables that were generated
network_model.x.pprint()
network_model.y.pprint()

x : Size=22, Index=i
    Key : Lower : Value : Upper : Fixed : Stale : Domain
      0 :     0 :  None :  None : False :  True : NonNegativeReals
      1 :     0 :  None :  None : False :  True : NonNegativeReals
      2 :     0 :  None :  None : False :  True : NonNegativeReals
      3 :     0 :  None :  None : False :  True : NonNegativeReals
      4 :     0 :  None :  None : False :  True : NonNegativeReals
      5 :     0 :  None :  None : False :  True : NonNegativeReals
      6 :     0 :  None :  None : False :  True : NonNegativeReals
      7 :     0 :  None :  None : False :  True : NonNegativeReals
      8 :     0 :  None :  None : False :  True : NonNegativeReals
      9 :     0 :  None :  None : False :  True : NonNegativeReals
     10 :     0 :  None :  None : False :  True : NonNegativeReals
     11 :     0 :  None :  None : False :  True : NonNegativeReals
     12 :     0 :  None :  None : False :  True : NonNegativeReals
     13 :     0 :  None :  None : False :  True : N

#### 2.4.3 Define the objective function

In [6]:
def n_objective(n_model: pyo.ConcreteModel) -> bool:
    """
    The objective function is given by the minimization of the sum
    of the unit crashing cost times the crashing time units over 
    all the model activities
    """
    
    return sum(n_model.costs[j]*n_model.y[j] for j in n_model.j)

# Add the objective function rule into the model
# Even though Pyomo by default performs minimization
# it is marked explicitly here
network_model.objective = pyo.Objective(rule=n_objective, sense=pyo.minimize)


# Print the objective function
network_model.objective.pprint()

objective : Size=1, Index=None, Active=True
    Key  : Active : Sense    : Expression
    None :   True : minimize : 166.66666666666666*y[0] + 500.0*y[1] + 214.28571428571428*y[2] + 125.0*y[3] + 50.0*y[4] + 62.5*y[5] + 25.0*y[6] + 400.0*y[7] + 100.0*y[8] + 181.8181818181818*y[9] + 194.44444444444446*y[10] + 66.66666666666667*y[11] + 80.0*y[12] + 100.0*y[13] + 33.333333333333336*y[14] + 52.94117647058823*y[15] + 33.333333333333336*y[16] + 150.0*y[17] + 50.0*y[18] + 58.333333333333336*y[19] + 156.25*y[20] + 250.0*y[21] + 28.571428571428573*y[22] + 25.0*y[23] + 500.0*y[24] + 166.66666666666666*y[25]


#### 2.4.4 Define the constraints

In [7]:
##########################
# CRASH TIME CONSTRAINTS #
##########################

def crash_time_constraints(n_model: pyo.ConcreteModel, j: int) -> bool:
    """Crash time constraints establish that the crash time
    for a given activity j must be lesser or equal to the difference
    between the normal and crashed times for that activity
    """
    
    return n_model.y[j] <= n_model.timediff[j]

# Add the crash time constraint into the model
network_model.crash_time = pyo.Constraint(network_model.j, rule=crash_time_constraints)
# Print the crash time constraints
network_model.crash_time.pprint()


crash_time : Size=28, Index=j, Active=True
    Key : Lower : Body  : Upper : Active
      0 :  -Inf :  y[0] :   3.0 :   True
      1 :  -Inf :  y[1] :   1.0 :   True
      2 :  -Inf :  y[2] :   7.0 :   True
      3 :  -Inf :  y[3] :   2.0 :   True
      4 :  -Inf :  y[4] :   5.0 :   True
      5 :  -Inf :  y[5] :   4.0 :   True
      6 :  -Inf :  y[6] :   4.0 :   True
      7 :  -Inf :  y[7] :   1.0 :   True
      8 :  -Inf :  y[8] :   2.0 :   True
      9 :  -Inf :  y[9] :  11.0 :   True
     10 :  -Inf : y[10] :   9.0 :   True
     11 :  -Inf : y[11] :   3.0 :   True
     12 :  -Inf : y[12] :   5.0 :   True
     13 :  -Inf : y[13] :   3.0 :   True
     14 :  -Inf : y[14] :   9.0 :   True
     15 :  -Inf : y[15] :  17.0 :   True
     16 :  -Inf : y[16] :   3.0 :   True
     17 :  -Inf : y[17] :   1.0 :   True
     18 :  -Inf : y[18] :   2.0 :   True
     19 :  -Inf : y[19] :  12.0 :   True
     20 :  -Inf : y[20] :  16.0 :   True
     21 :  -Inf : y[21] :  22.0 :   True
     22 :  -In

In [8]:
#######################
# NETWORK CONSTRAINTS #
#######################

def network_limit_constraints(n_model: pyo.ConcreteModel, a_dict: dict, constraintList: pyo.ConstraintList) -> None:
    """
    Network limit constraints establish that the time where a certain event i (node i)
    is reached (with respect with the initial even time of 0) is always greater or equal to
    the result of the normal duration of activity j that is related with the occurrence of 
    the event, minus the crashing time of the said activity j plus the occurrence time of
    the predecessor event k (node k) where the activity j originates from
    """
    # Rearanging the terms of the expression described above and using the
    # information provided by the problem dictionary
    for k, v in a_dict.items():
        constraintList.add(expr = n_model.x[v[1]] + n_model.y[k[1]] - n_model.x[v[0]] >= n_model.normaltime[k[1]])
    constraintList.pprint()
    return

In [9]:
# Create the constraint list
network_model.network_constraints = pyo.ConstraintList()
# Add the network inherent constraint list to the model
network_limit_constraints(network_model, activity_dict, network_model.network_constraints)

network_constraints : Size=28, Index=network_constraints_index, Active=True
    Key : Lower : Body                  : Upper : Active
      1 :   5.0 :    x[1] + y[0] - x[0] :  +Inf :   True
      2 :   2.0 :    x[2] + y[1] - x[1] :  +Inf :   True
      3 :  10.0 :    x[3] + y[2] - x[2] :  +Inf :   True
      4 :   3.0 :    x[4] + y[3] - x[1] :  +Inf :   True
      5 :   7.0 :    x[5] + y[4] - x[4] :  +Inf :   True
      6 :   0.0 :   x[5] + y[26] - x[3] :  +Inf :   True
      7 :   5.0 :    x[6] + y[5] - x[5] :  +Inf :   True
      8 :   6.0 :    x[7] + y[6] - x[5] :  +Inf :   True
      9 :   2.0 :    x[8] + y[7] - x[6] :  +Inf :   True
     10 :   3.0 :    x[9] + y[8] - x[7] :  +Inf :   True
     11 :  15.0 :   x[10] + y[9] - x[7] :  +Inf :   True
     12 :  12.0 :  x[11] + y[10] - x[8] :  +Inf :   True
     13 :   4.0 :  x[11] + y[11] - x[9] :  +Inf :   True
     14 :   8.0 : x[11] + y[12] - x[10] :  +Inf :   True
     15 :   5.0 : x[12] + y[13] - x[11] :  +Inf :   True
     16 :  1

In [10]:
##################################
# PROJECT COMPLETION CONSTRAINTS #
##################################
def project_completion_constraints(n_model: pyo.ConcreteModel, a_dict: dict, constraintList: pyo.ConstraintList) -> None:
    """
    Project completion constraints establish that the time occurence of the initial event
    is equal to 0 and that the conclusion date of the project must be lesser or equal
    than the deadline that is specified
    """
    # The start of the project is at time t=0
    constraintList.add(expr = n_model.x[0] == 0)
    # The project must have been completed by the deadline that is specified
    constraintList.add(expr = n_model.x[len(n_model.x)-1] <= DEADLINE)
    # We print these constraint after being added to the model
    constraintList.pprint()
    return

In [11]:
# We define the list for project completion constraints
network_model.completion_constraints = pyo.ConstraintList()

# We call the function project completion constraints
# and add these constraints to the  model
project_completion_constraints(network_model, activity_dict, network_model.completion_constraints)

completion_constraints : Size=2, Index=completion_constraints_index, Active=True
    Key : Lower : Body  : Upper : Active
      1 :   0.0 :  x[0] :   0.0 :   True
      2 :  -Inf : x[21] :  60.0 :   True


#### 2.4.5 Solve the linear programming model

In [12]:
# We import using pyo.Suffix to later
# perform a sensitivities analysis
network_model.dual = pyo.Suffix(direction=pyo.Suffix.IMPORT)

# We use the LP Gurobi solver
# to find the optimal solution
Solver = SolverFactory('gurobi')

# Solve the model using the solver
# Start the stopwatch
clock_start = process_time()
Results = Solver.solve(network_model)
clock_stop = process_time()

In [13]:
# Display the solution for problem
print(f"Elapsed time: {clock_stop-clock_start}")
network_model.display()

Elapsed time: 0.046875
Model unknown

  Variables:
    x : Size=22, Index=i
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   0.0 :  None : False : False : NonNegativeReals
          1 :     0 :   2.0 :  None : False : False : NonNegativeReals
          2 :     0 :   4.0 :  None : False : False : NonNegativeReals
          3 :     0 :  12.0 :  None : False : False : NonNegativeReals
          4 :     0 :   5.0 :  None : False : False : NonNegativeReals
          5 :     0 :  12.0 :  None : False : False : NonNegativeReals
          6 :     0 :  13.0 :  None : False : False : NonNegativeReals
          7 :     0 :  14.0 :  None : False : False : NonNegativeReals
          8 :     0 :  15.0 :  None : False : False : NonNegativeReals
          9 :     0 :  23.0 :  None : False : False : NonNegativeReals
         10 :     0 :  24.0 :  None : False : False : NonNegativeReals
         11 :     0 :  27.0 :  None : False : False : NonNegativeReals
         12

#### 2.4.6 Interpret the network linear programming model results



In this case, we have 22 events and 28 activities that create an incidence matrix with 616 entries. We then have 22 $x_i$ variables and 28 $y_j$ variables that were obtained by solving the problem.

As previously stated we can without any issue (due to the aforecited theorem) allow for real solutions as an integer input will always lead to an integer output.

The solutions for $x_i$ mean that for each $x_i = s_i$, where $s$ is the integer solution, that event $x_i$ takes place at the moment $s$, which represents the timely departure from the project beginning at $x_0 = 0 \text{ days}$. We can check that the last activity, $x_{21} = 60 \text{ days}$ which is within the deadline that was specified to the model. Moreover, we can check that all $x_i$ are higher or equal to 0 and lesser or equal than 60.

For $y_j$ we have the amount of days that the model finds to be optimal to be reduce for each activity *for the model to be completed within schedule*. For instance activity $x_5 = 4 \text{ days}$ means that, by a increasing the cost at that stage, we can reduce (crash) its duration by 4 days; the same applies for the remaining $y_j$ values.

The objective function allows us to see the crashing cost that we incurred in so as to ensure that the project is completed within schedule.

#### 2.4.7 Sensitivies and their interpretation

In [14]:
# After importing the corresponding suffix
# in the previous cell, we display the
# duals/sensitivities per constraint
network_model.dual.display()

dual : Direction=Suffix.IMPORT, Datatype=Suffix.FLOAT
    Key                       : Value
    completion_constraints[1] :                 0.0
    completion_constraints[2] :  -244.3181818181818
                crash_time[0] :  -77.65151515151516
               crash_time[10] :                 0.0
               crash_time[11] :                 0.0
               crash_time[12] : -101.81818181818181
               crash_time[13] :                 0.0
               crash_time[14] : -158.04367201426024
               crash_time[15] :                 0.0
               crash_time[16] :                 0.0
               crash_time[17] :  -41.37700534759358
               crash_time[18] :                 0.0
               crash_time[19] :                 0.0
                crash_time[1] :                 0.0
               crash_time[20] :                 0.0
               crash_time[21] :                 0.0
               crash_time[22] :                 0.0
               crash_tim

Recall that these sensitivies express how much the objective function would vary if a given variable linked to a constraint were to have a unit change.

We can see that the first completion constraint $x_0 = 0.0$ and rightfully so, as it is clear that this inequality needs to hold for the model to be meaningful. Now, for $x_{A-1} = -244.3\overline{18}$ makes sense: one day more would entail less crashing costs, as we have more time to complete the project, and that is why the objective function (the total crashing cost) would diminish.